Message format for RadarTrack

radarAaron opened the pull request:
Proposed message format for RadarTrack

Prior to proposing this format, research was done into the radar data message formats that exist out in the wild. The details of this analysis can be found here: https://github.com/radarAaron/radar_msgs/blob/master/ROS%20Message%20format%20comparison.xlsx

## Summary
**Position in 3D Space**
All position appears to be represented in Cartesian coordinates, there is no consensus as to the format however or how to represent uncertainty.

The following formats were used:
* geometry_msgs/Polygon (no uncertainty)
* geometry_msgs/PoseWithCovariance
* x,y coordinates [also has a second message type which use distance, bearing] (no uncertainty)
* x,y,z (no uncertainty)

**Velocity**
There is no standard for velocity. All formats could be converted to Vector3 however.

The following formats were used:
* geometry_msgs/Vector3 (no uncertainty)
* geometry_msgs/TwistWithCovariance (covariance)
* x,y velocities (no uncertainty)
* float32 speed (no uncertainty)

**Acceleration**
Two drivers output acceleration.
* geometry_msgs/Vector3 linear_acceleration 
* geometry_msgs/AccelWithCovariance

**Size**
There is no standardization on object size.
* geometry_msgs/Polygon (no uncertainty)
* length, width  (no uncertainty)
* Horizontal and vertical cross section (no uncertainty)

**Classification**
Only one format outputs any form of classification

**Additional fields**
The following fields are also present in some formats: orientation_angle, msg counter, target_format_type, quality, speed_trigger_flag, Lane, Track_packet_counter

**Conclusion**
* Using the Cartesian coordinates system appears to be most standard. 

* Velocity could be reconciled into a Vector3 type with the velocity components which are not supported by a driver being set to zero.

* Acceleration is probably best represented as a Vector3

There is no consensus about how the size of a shape is reported, with the only option being RCS. The call was made to represent 'size' using an ovaloid shape based on the distribution of points in 3D space.

* Only one driver outputs classification but this is still probably a useful addition to the standard format.

* It is unclear how uncertainty should be presented

## Notes
Object tracks are based on usually based on a point cloud. The point cloud shape is somewhat object specific and does not necessarily cover the full size of the object (eg. points tend to concentrate towards the center). The 'size' format chosen here is based on the distribution of points in 3D space. An **alternative** approach for expressing the size would be as a covariance matrix.

## Questions
* Should size be expressed as 'size' or as a covariance matrix?
* Is expressing the uncertainty of the velocity and acceleration components important?

aaronfultonnz committed:
Added message format for RadarTrack

tfoote commented:
> Should size be expressed as 'size' or as a covariance matrix?

What is the underlying measurement that's trying to be expressed here? If it's a clustering of points then a size vector of the bounding box makes sense. The alignment of the bounding box should be explicit too. Covariance is semantically very different.

> Is expressing the uncertainty of the velocity and acceleration components important?

Uncertainty is probably worth expressing. Since these are the results of an estimator there should exist some estimates of the uncertainty. 

In the recent message API review it was determined that it would be valuable to improve our support for covariance based uncertainty. https://github.com/ros2/common_interfaces/blob/master/docs/APIReviewFoxy.md A new 3x3 Upper Triangle covariance representation would go well here for these measurements.

tfoote reviewed (dismissed):
Without the header the RadarTrack isn't really useful on it's own so the RadarTracks is a little bit odd that the smallest unit you can subscribe to would be plural. This is presumably some collection of tracks segregated by something like time. I'd suggest considering if there's a better name but could be convinced otherwise.

tfoote commented on msg/RadarTrack.msg:
```diff
@@ -0,0 +1,12 @@
+                                            # NB: The z component of these fields is ignored for 2D tracking.
+geometry_msgs/Point position                # x, y, z coordinates of the centroid of the object being tracked.
+geometry_msgs/Vector3 velocity              # The velocity of the object in each spacial dimension.
+geometry_msgs/Vector3 acceleration          # The acceleration of the object in each spacial dimension.
```
After reading through #2 again this looks a lot like an annotated Detection with track_id, acceleration and classification added. That relationship might be valuable to leverage with a sub datatype.

radarAaron replied:
The call has subsequently been made to do away with the Detection type entirely.

peci1 commented on msg/RadarTrack.msg:
```diff
@@ -0,0 +1,12 @@
+
+                                            # NB: The z component of these fields is ignored for 2D tracking.
+geometry_msgs/Point position                # x, y, z coordinates of the centroid of the object being tracked.
+geometry_msgs/Vector3 velocity              # The velocity of the object in each spacial dimension.
```
spatial is the more common spelling: https://www.merriam-webster.com/dictionary/spatial

peci1 commented on msg/RadarTrack.msg:
```diff
@@ -0,0 +1,12 @@
+                                            # NB: The z component of these fields is ignored for 2D tracking.
+geometry_msgs/Point position                # x, y, z coordinates of the centroid of the object being tracked.
+geometry_msgs/Vector3 velocity              # The velocity of the object in each spacial dimension.
+geometry_msgs/Vector3 acceleration          # The acceleration of the object in each spacial dimension.
```
spatial

RonaldEnsing commented:
How radar specific is this message type? What about a generic (sensor agnostic) track msg?

icolwell-as commented on msg/RadarTrack.msg:
```diff
@@ -0,0 +1,12 @@
+# All variables below are relative to the radar's frame of reference.
+# This message is not meant to be used alone but as part of a stamped or array message.
+
+uint16 track_id                             # The track ID of the object generated by the radar.
```
Since the name of the message is "RadarTrack", would it be simpler/intuitive to use `id` instead of `track_id`?

SteveMacenski replied:
I like the use of the unique_msgs/UUID and just calling it a uuid since that's clear

icolwell-as replied:
Does UUID imply a certain format though?  Not sure if all radars would provide the data in the UUID format, or is that something the drivers would have to handle?

SteveMacenski replied:
This is just an id generator tool, so anything should be able to generate a UUID. Its more or less just a long string of numbers like a big uint16 but its a standard way of dealing with data like this

radarAaron replied:
My preference is for ID over UUID. UUID's are 128 bits long (uint8[16] as per http://docs.ros.org/api/uuid_msgs/html/msg/UniqueID.html). UUID seems like a complete overkill for a numbering system that will seldomly go above 100 and usually a lot less. UUID also implies numbers will be universally unique. I suspect most drivers will output an integer ID and the numbering will reset when the module does. ID at least does not imply the numbers are universal unique

icolwell-as replied:
This conversation was marked resolved, but it seems like it was resolved the wrong way? Sounds like we agreed to use a simple integer ID but a UUID was committed.

To clarify the discussion in this thread. I believe @radarAaron meant it would seldomly go above 100 *for a single frame*, meaning the radar is rarely tracking more that 100 objects at one instant in time. From my limited automotive radar experience, tracked objects are usually given an 8-bit integer ID.
I'm not against UUID, it's definitely handy, but it's not something that any radar hardware (that I've used) would provide. I think for the majority of radar hardware, we'd have to sort of fill in this UUID from the ROS driver side.

peci1 replied:
I haven't considered this conversation as resolved (although it has been marked so, maybe automatically by a commit changing the line?). 

Again, there seems to be some confusion which might have been removed if there were a better description of what this message actually is and should be. @icolwell-as is talking about IDs unique among all tracked objects in a single time instant, while @SteveMacenski is talking about IDs aggregated over days of operation.

radarAaron replied:
I had understood that UUID was better from an ROS/scalability/multi-sensor perspective and committed on that basis. I agree with @icolwell-as that we would have to fill this out driver-side as the hardware would generally would supply an 8bit ID, but that is not too difficult to do.

icolwell-as commented on msg/RadarTrack.msg:
```diff
@@ -0,0 +1,12 @@
+geometry_msgs/Vector3 acceleration          # The acceleration of the object in each spacial dimension.
+geometry_msgs/Vector3 size                  # The average deviation (m) in the x, y, z, dimensions
+                                            #   between the center of the object and the data points which it is derived from.
+string classification                       # An optional classification of the object (eg. Person, Car).
```
I second this. In my limited radar experience, I've only ever seen things classified by "movement". Meaning an object is classified as either "stationary/static" or "moving/dynamic".

radarAaron replied:
It is possible to make some determination about what type of object is being detected based on the signature of the data and some drivers could/do implement this. The AUTOLIV 77GHz FLR RADAR driver has 'object_type' for example. I think it would be good to allow for this valuable information to be populated if it is available.

In regards to an ENUM. As far as I know there is no such list we could draw from. I very much expect the classification will be manufacturer and application specific. Using string allows the field to be used in an application specific way, while still maintaining the common message format. I think this is a good compromise.

radarAaron replied:
OK that sounds like a reasonable compromise.

aaronfultonnz committed:
Spelling correction

gavanderhoorn commented:
I'm only an interested external observer (with no experience with radar or radars), but would it be helpful to design these messages with a bit more of a "future radar" in mind, next to the devices (and drivers) which exist today and which have been investigated?

Future devices might be significantly improved, making this:

> UUID seems like a complete overkill for a numbering system that will seldomly go above 100 and usually a lot less.

a statement which may prove to become untrue in the (near?) future (obligatory reference to "[640KB ought to be enough for everybody](https://en.wikiquote.org/wiki/Talk:Bill_Gates)").

> UUID also implies numbers will be universally unique. I suspect most drivers will output an integer ID and the numbering will reset when the module does. ID at least does not imply the numbers are universal unique

re: drivers have internal IDs which are not globally unique: this is most likely true, but there could still be value in using a UUID as it would avoid conflicting IDs *between* drivers (both multiple instances of the same driver/sensor and multiple different sensors each with their own driver.

A similar discussion took place over in `vision_msgs`: https://github.com/Kukanani/vision_msgs/issues/25, https://github.com/Kukanani/vision_msgs/issues/18 and https://github.com/Kukanani/vision_msgs/pull/19.

The question there was whether a globally unique ID should be included in all messages (also detections), or whether a `string` should be used instead.

As @RonaldEnsing writes above, there seem to be quite some parallels between "detections" here and in `vision_msgs`. Perhaps the discussion there could be helpful here.

SteveMacenski commented:
> How radar specific is this message type? What about a generic (sensor agnostic) track msg?

@RonaldEnsing agreed, this could at some point be moved into a more generic tracking messages. You're about 2 steps ahead of us. Vision messages is vision specific and unfortunately some of the assumptions they made don't really fit outside of pointcloud processing for 3D items. We'll likely make a `detection_msgs` or `tracking_msgs` package that handles generic detections/tracks based on the lessons learned from this. We also need to suppose more generic detection types in navigation2, but we're a bit away from that still. First step is radar, second step is everything else.  

> between drivers (both multiple instances of the same driver/sensor and multiple different sensors each with their own driver).

Excellent point, in fact that would only mean UUID is the way to go because otherwise you can't reasonably promise uniqueness of the tracks.

aaronfultonnz committed:
Altered classification to uint8

aaronfultonnz committed:
Changed track_id to UUID and updated the description of 'size'

aaronfultonnz committed:
Updated description of size to allow for better compatiablity of rectangle-shaped bounding boxes.

radarAaron commented:
@SteveMacenski I'm happy to merge this request as is. Any last suggestions before I do?

SteveMacenski commented:
@peci1 @tfoote @gavanderhoorn Can you review again and let us know your thoughts / approvals?

We need to get general buy in @radarAaron so we need to have multiple approvals to merge

peci1 commented:
I generally agree with all unresolved @tfoote's comments.

Utilizing RadarDetection type here and only adding id, acceleration and classification would make sense to me. Or, rather check my understanding: the track is created by a process that takes a detection, finds its corresponding track, and updates the track with data from the new detection. This basically creates something that could be described as "the detection that best suits the observations", or "if I had an oracle, I'd see the object exactly as this detection". The use of `RadarDetection` would help making this clearer from the Track definition. On the other hand, it might (wrongly) imply to someone that the exact values stored in the `RadarDetection` subfield are some actual data measured by the radar at some time (which would be wrong because the data are the result of a tracker/estimator).

I vote for adding covariances everywhere where it makes sense - as @gavanderhoorn stated, we should look into the future. If some vendor doesn't want to compute this information, let's allow them to publish an invalid (zero) covariance. It should be explicitly stated in a comment that downstream code should be prepared for the case of zero covariances.

The `size` field is still pretty vague. Is it an axis-aligned bounding box? Or is it an oriented bounding box? But that would require an orientation field, and it should also specify whether it is a minimal volume OBB or of its main axes are determined by PCA or whatever... Or is it an oriented bounding box that's always facing the radar frame? The last case would probably make sense given the method how radars work, but then the size would change as the object would be moving around the radar... I'd vote for the PCA-derived OBB or bounding ellipsoid.

I see that the previous paragraph about the size field is quite a stream of thoughts, and I hope you'll be able to extract ideas from it. Maybe a good start would be to agree on the desired properties of the size. Should it remain unchanged when the object/radar moves around (given that the object has already been seen from all sides so that its 3D size can be correctly estimated)? Or is it okay if it "fluctuates", e.g. the way that happens when you rotate an irregular object and observe its axis-aligned bounding box? And, again, even the size field should implement some kind of uncertainty or covariance (whatever it means for boxes). To bring in more complications, an OBB can have uncertainty in its dimensions, orientation and position, and all three of them are probably correlated...

Another thought is about the position, velocity and acceleration fields - they're all missing the rotational components. Going back to @gavanderhoorn's comment, isn't it (theoretically) possible that some future radars would be able to return e.g. rotational velocity of the object (and thus its rotation and angular acceleration would also be available)? Again, not all radars would have to support it - all unsupported "fields" would be set to zero.

And the very last thought is inspired by the camera info messages. It is apparent that each radar (and its driver) has a different set of capabilities. I think it might be useful for downstream code if it could determine the "capabilities" of the driver. For me, it'd be important to know if the acceleration is zero or unobservable. So I suggest creating a `RadarInfo` message that could be published latched on a defined topic and would apply to all radar track messages published after this info message (this would be inconsistent with `CameraInfo`, but it seems more appropriate to me here as the capabilities would be constant opposed to camera intrinsics which can change over time). This message would describe the capabilities using boolean flags, e.g. `acceleration_available`, or even more granular `acceleration_x_available` etc. It could also contain some textual descriptions of the `classification` values, which would nicely bridge it back to the originally proposed `string` representation. Alternatively, no info message would have to be created if there were a convention that all "unavailable" fields would be set to NaN. I know working with NaNs is tricky, but at least you'd see that your computations rely on unobserved values.

SteveMacenski commented:
That's a long post - can you submit a specific recommended change in the form of a suggested message?

peci1 commented:
As I don't work with radars, I don't know the capabilities they have now or could gain in the future, and thus I can't decide whether my requests are worth it or not. If all of them were valid, we'd end up with something like:

RadarDetection:
```
geometry_msgs/PoseWithCovariance position
geometry_msgs/TwistWithCovariance velocity
geometry_msgs/Vector3 bounding_box_size
float64[9] bounding_box_size_uncertainty  # something like covariance
geometry_msgs/Pose bounding_box_pose
```

RadarTrack:
```
uuid_msgs/UniqueID uuid
radar_msgs/RadarDetection detection
geometry_msgs/AccelWithCovariance acceleration
uint8 classification
```

RadarCapabilities:
```
string frame_id  # to specify which radar does it describe? maybe not needed?
bool three_dimensional # false would automatically "disable" all 3rd dimension fields
bool orientation_available
bool angular_velocity_available
bool linear_acceleration_available
bool angular_acceleration_available
... # mayb more
string[] classification_categories # index in list corresponds to classification value
```

But maybe would, in the end, need different definition of `size` for detection (2D) and for track (2D or 3D), and that would prevent using `RadarDetection` as the subfield.

Also, for radars working only in 2D, the orientation as a quaternion would not make sense...

This suggestion needs more thinking and polishing, I just wanted to show the ideas. I'm not insisting on any of them (but not adding covariances would need more convincing).

peci1 commented:
Maybe the whole Radar* API could be duplicated - once with covariances, once without them. As is `Pose` and `PoseWithCovariance`. That would allow vendors to start with simple and smaller messages and as they progress with development, covariances could be added still using a standard message. It's a pity that ROS 1 doesn't support memory-efficient optional message fields...

Also, I think it's insufficient to review the implementations of current ROS drivers to determine whether covariances are widely used or not. For that, it seems to me more suitable to look at the low-level drivers of the radars and look there if some covariances can be read out or not (because ROS drivers are often just wrappers around some custom low-level drivers, and the low-level driver might be much more versatile than the ROS wrapper).

SteveMacenski commented:
I don't think anyone except @radarAaron is arguing to not have covariances. I think they should be outright added and we can move on with this PR .

radarAaron commented:
@peci1 Radar does not provide a high resolution of information. I don't think it will ever be possible to get the rotational velocity etc out of radar data. Radar can identify where there is a high probability of an object and its rough size and direction. The data is more of a point cloud of data which you have to make inferences about.

In terms of the size and coordinates: These is intended to the frame of reference of the radar sensor. Oriented Bounding Box using something like Principal Component Analysis is really something that makes a lot of sense in terms of radar data, the data is just not suited to that kind of analysis, and is therefore not something that should be part of a common message format. The RadarPoint message format is available to anyone who wants to try and implement such methods.

This RadarTrack message will come out of some higher-level algorithm on the radar hardware (something like a Kalman filter with k-means clustering) but that algorithm will be vendor specific and I expect there to be a lot of variation in what each piece of hardware is capable of, what the vendors implement, and what makes sense for a specific application, to the point where I think the idiosyncrasies are best explained the modules manual and that a trying to condense these down into a standardized message not really doable.

aaronfultonnz committed:
Added covariance matrix

radarAaron commented:
@SteveMacenski I've added a covariance field to the message type. Does what I've added make sense from an ROS perspective? If not can you suggest an alternative. As we have mentioned previously, I certainly wouldn't expect this to be populated by many vendors.

SteveMacenski commented:
See IMU message: http://docs.ros.org/melodic/api/sensor_msgs/html/msg/Imu.html I think each of the covariances should be separated by quantity measured (e.g. a few 9 entry matrices not 1 big one. Also more compact)

aaronfultonnz committed:
Adjusted covariance format

radarAaron commented:
Thanks @SteveMacenski. I had used the PoseWithCovariance.msg as my reference (which only had one covariance field). Covariance fields are now adjusted.

JWhitleyWork commented:
This discussion seems to have died down a few months ago. However, it looks like all issues with this particular message were resolved. @SteveMacenski - would you be willing to merge this?

SteveMacenski commented:
@peci1 @tfoote any blocking suggestions?  I want to make sure we have near-universal 'OK' on this. I'm 100% fine iterating on this over time, but so far from Aaron's expertise as the CEO of a radar company, it seems like this represents the viable set of things a radar can provide and would be able to replace all the radar message uses we highlighted in the survey results from the first comment.  

---

I agree with @peci1's sentiment that `XYZWithCovariance` would be nice to use, but none of those provided make sense in this context. If radar's can't meaningfully provide angular velocities, Twist is not right. If pose isn't giving orientation Pose is not right. Ultimately, we have precedence for doing separated covariances from IMU messages (and probably others).

peci1 commented:
I see a few unresolved/undismissed ideas.

- [ ] Include a RadarDetection as part of this message?
- [ ] The `size` field documentation is still insufficient. It'd be better if it described what kind of dimension reading it is (AABB, OBB, RCS, fitted ellipsoid) and regarding which reference frame (e.g. an AABB can be aligned either to the message's reference frame, or to the viewing direction to its centroid). The parenthesized comment `(eg Radius of half-width/length/height)` doesn't help much. So is it a size or a half-size?
- [ ] Value `0` in `classification` should mean no classification was performed. static and dynamic should then be 1 and 2. Maybe create named constants for these three significant values?
- [ ] Should unobservable components be required to hold a NAN instead of zero? I know it's more convenient to require zeros as they are the default value, but having a way to determine observed zero from an unobserved value could still be handy. It seems to me that if some driver vendor mistakenly sets 0 instead of a NAN, the worst thing that could happen is that some estimator would be over-confident.

radarAaron commented:
@peci1 I agree with your comments
1. Radar Detection (See my thoughts on the RadarDetection message pull request)
2. How is this for a better size description: "The object radius as a fitted ellipsoid (m) in the x, y, z, dimensions". The comment at the top of the message format already says "All variables below are relative to the radar's frame of reference."
3. Agreed that 0 should mean no classification
4. I agree that NAN is clearer than 0 if a field is not supported. is ".NAN" the correct nomenclature for ROS message formats?

peci1 commented:
@radarAaron Okay, the new wording for `size` is good. Regarding NAN, I don't think there is a ROS-specific representation, as ROS is language-agnostic. However, if you want to pass NAN from commandline to e.g. `rostopic pub`, the syntax is `.NaN` (that is given by the YAML standard; or maybe is it case insensitive?). So either this, or the generally used term `NaN` (I use NAN just out of laziness).

SteveMacenski commented:
1. I think we're discussing removing RadarDetection message completely, so that might be a moot point. Lets see what Josh says

2. Please suggest an alternative description of the semantic meaning of `size`, but that description should not require a specific method of compute. Each radar provider might compute it differently, but the meaning of `size` shouldn't relate to the specific algorithm used to define it (e.g. an ellipsoid based on all points in cluster vs ellipsoid based on probabilistic values derived from points in the cluster are both value implementations of size with different covriances. Even different algorithms are going to result in slightly different outcomes, but nothing in any ROS message should be requiring a specific methodology, that doesn't make sense). The reference frame comment is important, I agree. 

3.  Agreed. 0 should be unknown

4. That isn't consistent with any other ROS message and for no other reason I think that we should not specify that. It is true that different sensors might implement or not implement different parts of this message, but that's always been true with pointcloud, IMU, and other messages. There's nothing unique about radars in that regard and the community has been doing fine in light of that with many sensors for years. Not to say that it couldn't be improved by NaNs in this regard, but that's a larger problem that we should potentially file an `REP` for creating that standard across all sensor messages of this style.

JWhitleyWork commented:
1. I per my comment on #2 - I think just Track and Return are good enough.
2. I agree - a definition for what's included in the message field is important but how it was generated and whether or not it is exactly what the radar provided is not important. Standardization is.
3. Agreed.
4. What about flags outside of the fields themselves to indicate used fields? Or, for example, different supported fields based on a type enum (ala [shape_msgs/SolidPrimitive](http://docs.ros.org/en/api/shape_msgs/html/msg/SolidPrimitive.html)?

aaronfultonnz committed:
Altered notes for classification and size fields as per feedback on the pull request

radarAaron commented:
I've pushed an update to the message format which includes the change in classification enum and a new description for size "The object size as represented by the radar sensor eg. length, width, height OR the diameter of an ellipsoid in the x, y, z, dimensions" Does this sound OK. If not can you please suggest other wording. I used the two examples to try and explain what is expected in the size field and by choosing diameter as opposed to radius for the ellipsoid example to be a closer match to L x W x H for greater consistency.

4. @SteveMacenski what do you think about flags to indicate the used fields? Is this common practice with other ROS messages?

SteveMacenski commented:
1 and 3 sounds like we're done

2. I'm OK with that as long as no one interprets it as recommending one method of computation of the size over another. I think the LWH / diameter description is ok. I don't think it makes it clear the frame of reference though of the size. Is this in the object's XYZ size or the sensor frame's view of the XYZ size? I assume its in the sensor frame since we cannot know the object orientation. That should be explicit. 

4. I've typically only seen that in visualization-based messages. I don't think that makes things any better for this application.

aaronfultonnz committed:
Making the size frame of reference explicit

radarAaron commented:
2. I've added the frame of reference to the description
3. I agree that we don't include any addition parameters designating which fields are used.

SteveMacenski commented:
@radarAaron @JWhitleyWork @peci1 @tfoote  are you happy with the state of this? If so, submit an approval review, if not, please highlight any un-resolved issues we can address. It seems like we have the momentum to get this across the finish line right now and I'd love to capitalize on it. 

if you 4 approve, I'll put 1 more "FYI" discourse post on it. If no more meaningful objections, we'll get this in before the end of november. I'll commit to release and setting up CI.

peci1 commented:
2. I'm still not sure about the reference frame specification. Thinking it through more deeply, I've found out that it is specified nowhere in this message whether radar frame is the actual frame of the "solid/static" radar housing, or if it is the frame that's revolving around with the emitter. If the first case was true, then I think the size frame wouldn't be the radar frame but the sweeping emitter frame (which may or may not be represented explicitly). What are your thoughts on this? I think that it might not be needed to specify which of the two types the radar frame is (both are correct if you transform the values correctly), but the size field needs to work the same in both cases.

4. Okay, if you feel NANs are not a good way to go, then we can use zeros. If some code would really want to detect which values are unobserved, it can probably deduce it from covariance - I'd treat exact zero covariance as a sign that the component is not measured and uses a placeholder constant instead.

SteveMacenski commented:
@peci1 There's almost always going to be some static frame for adding the sensor to a robot UDRF / transformations file. If that, however, was not the case and was just from the rotating emitter, what difference would that make? The `RadarTracks` header has a `frame_id`. If it was a rotating frame, we still have the `frame_id` to transform at that `stamp` with TF into our `base_link` frame. If this rotating sensor frame doesn't update TF then the issue of frames are moot because its immeasureable or unusable to the larger robot system. Nothing about this interface cares if `frame_id` is static relative to the robot or rotating.

peci1 commented:
@SteveMacenski I think the `size` field cares. If I understand it correctly (and correct me if I'm wrong), the size most radars will report will be perpendicular to the "ray" or "rays" that measured the object. Which will be different in case frame_id points to the static frame. And I thought the consensus here was that it is not desirable to re-compute the size somehow (transform it to the static frame) just for the sake of filling up a field in the message.

SteveMacenski commented:
that is no different of a concern for velocity, acceleration, and position as well; size is not unique there. Again, the header's frame would be populated at a timestamp of the rotating frame and the results can be backed out. For a rotating lidar that is returning some assemblance of tracks, it itself will need to buffer the rotating information into some global frame in order to cluster and produce tracks. That is then likely the frame that this information would then be published in. I don't see an issue.

peci1 commented:
For vector- or point-style fields, the way of transformation between frames is obvious, no doubt about that. But how would you transform an axis-aligned ellipsoid?

I tried to chart a situation. The radar sees the light blue object and returns a size measurement that corresponds to the red ellipsoid (I know that if this were a top-down view of a real radar, we wouldn't know the depth of occluded parts, but let's keep it there for the sake of explanation). How would you transform it to the axis-aligned frame of the radar denoted by the green-red axes in the middle? Smallest circumscribed axis-aligned ellipsoid? And is that what people really want? Wouldn't it be easier to say that the size ellipsoid (or any other representation) is always aligned to a frame given by the radar frame position and the estimate of track position? In case of revolving frame, it would make no change, but it would be a change for the case of static radar frame. Any thoughts, @radarAaron ?

![obrazek](https://user-images.githubusercontent.com/182533/99587338-41069000-29e9-11eb-9f2b-1396d8d9ab5c.png)

SteveMacenski commented:
I think the situation you bring up is conceptually flawed. Your rotating radar will take measurements of the style `RadarReturns`. The radar will then buffer those returns into its internal frame so that it can combine all of its data which then is clustered for producing a detection and then tracking. So I think that means the concept of the blue detection is flawed since it wouldn't be done at that stage, or combined with other measurements to create discrete detections.  

Separate of that argument, applying a homogeneous transformation matrix to any scalar or vector is well described and understood. If you have dimensions of ellipsoid in XYZ relative to frame `A`, converting them into a colinear frame (or any other frame for that matter) is a matter of algebra (or TF2).

peci1 commented:
I still think our thought lines are a bit misaligned.

Before we continue the discussion, can we agree that transforming an axis-aligned (or frame-aligned, to stay with the terminology) object to another frame is not just about multiplying its dimensions vector by a transformation matrix? Imagine an axis-aligned box. Now transform its corners to a frame rotated 45 degrees. The dimensions in the local frame of the box did not change, but the box lost the property of being axis-aligned in the new frame. So you can't even say what it's dimensions are in the new frame, because it became a different class of object (and you can only measure axis-aligned boxes for the sake of this example). If you want it to retain the axis-aligned box property, you have to do "something", but that "something" is not really defined in a unique way.

radarAaron commented:
I'm struggling to understand all the nuances of the conversation here. The 'size' of a radar object is only ever going to be a fairly rough approximation and the size and shape will quite probably vary between frames as the object moves or rotates. 
@peci1 I understand your analogy of the box. Firstly, from the radar's perspective, there will never be enough resolution to determine the object is a box in the first place - everything just looks like blobs to a radar. The radar will always use its coordinate system for all measurements its takes.  I'm not sure if my contribution has helped clear things up.

peci1 commented:
I used the analogy with a box just to illustrate the geometrical problem. Otherwise, we're talking about ellipsoids here, right?

I think I still need a little bit more clarification. If the radar works in polar coordinates (which a revolving radar would), I'd expect the measured blobs (or their estimated ellipsoids) to have axes tangent to the circle that is centered at the radar and goes through the blob's centroid. Especially, the estimated ellipsoids would not be aligned to the fixed Cartesian radar frame (except for those at azimuths 0, 90, 180 and 270 degrees). Now, if we sticked with the current meaning of the `size` field, it would mean they would have to be "somehow" transformed to ellipsoids aligned to the fixed frame. The question is whether such transformation is intended or useful. I would choose to leave there the "raw" estimate of the radar, which would be an ellipsoid not aligned to frame_id. Or, do some radars output size estimates as ellipsoids aligned to a fixed cartesian frame? (i.e. no rotation, axes of the ellipsoid parallel to the frame axes).

radarAaron commented:
> Or, do some radars output size estimates as ellipsoids aligned to a fixed cartesian frame? (i.e. no rotation, axes of the ellipsoid parallel to the frame axes).

FMCW radar (which is is the type RadarIQ uses and is also the type used in vehicles) is a fixed position and does not rotate. These sensors have about a 100 degree horizontal field of view, which can be all be seen simultaneously. If you use a filter to hide static objects, you could think of it as a camera which only sees moving objects. The 'raw' radar returns use polar coordinates and looks like a sparse point cloud (`RadarReturn`). When clustering algorithms/EKF's etc are applied these can be turn into 'objects', which are usually expressed in a Cartesian frame. It is this type of Radar I had in mind when creating the message type.

Rotating radar sensors (as you might find on a ship) are a whole different thing and would probably warrant their own message type.

peci1 commented:
Now we're home :) I thought this message should also cover the revolving radars. For FMCW, the discussion about the frame of `size` is pointless.

So it probably deserves clarification in the message definition.

aaronfultonnz committed:
Added comment about that this message relating only to FMCW radar

radarAaron commented:
Very good idea to add a comment about FMCW radar!
If we can learn one thing from our discussions it is that having some very clear documentation about radar, how it works, what is possible and what is not as how the message are supposed to function will be very valuable.

peci1 commented:
>  having some very clear documentation about radar

Agree!

One last point from my side - what about the named constants for the three pre-defined classification types?

aaronfultonnz committed:
Added constants for object classification field

radarAaron commented:
Named constants added

peci1 reviewed (approved):
Perfect! Thanks for patience.

SteveMacenski commented:
Are the static / dynamic typical enum information a radar SDK / sensor would provide to populate those specific items? Settings those in the message will standardized those values for the message so we need to be clear that this is not just an example, but creating a standard.

JWhitleyWork reviewed (approved):
LGTM

JWhitleyWork commented:
@tfoote I think you're the last set of eyes.

icolwell-as commented on msg/RadarTrack.msg:
```diff
@@ -0,0 +1,23 @@
+float64[9] position_covariance              # Row-major about the x, y, z axes
+float64[9] velocity_covariance              # Row-major about the x, y, z axes
+float64[9] acceleration_covariance          # Row-major about the x, y, z axes
+float64[9] size_covariance                  # Row-major about the x, y, z axes
```
Very minor detail but I just noticed the github warning about a missing newline at the end of the file. Maybe not relevant for *.msg files?

radarAaron replied:
New line added

aaronfultonnz committed:
Added new line to the end of the file

tfoote commented on msg/RadarTrack.msg:
```diff
@@ -0,0 +1,23 @@
+# All variables below are relative to the radar's frame of reference.
+# This message is not meant to be used alone but as part of a stamped or array message.
+
+# Object classifications (Additional vendor-specific classifications are permitted eg. 3 = Car)
```
I generally don't like leaving these open. But definitely don't suggest the usage of 3. This should be expanded to maybe a uint16 and reserve the upper half for additional vendor specific classifications and leave the lower half for potential future standardization. 

On the scale of this message optimizing 8 bit vs 16 bit is not worth it.

SteveMacenski replied:
Agreed. @radarAaron I think the action items here are:
- change uint8 to uint16 for all pre-defined classifications & `classification`.
- Keep `car` but remove `3 =` since that implicitly sets 3 as a car, which I don't think was the intent. 

I would suggest we could reserve the first `100` for further standardization rather than 32,000 - that seems excessive. Thoughts?

gavanderhoorn replied:
> I would suggest we could reserve the first `100` for further standardization rather than 32,000 - that seems excessive.

In my experience you end up never having enough of anything. Since you're already going to `uint16`, *really* future-proof it and do what @tfoote suggests.

Users won't care whether OEM ids start at 100 or 32000, it's just a number.

JWhitleyWork replied:
I agree with @gavanderhoorn - Let's start vendor-specific values soemwhere around 32000.

radarAaron replied:
Changes made as requested.

SteveMacenski replied:
@radarAaron one last part of that - you need to update the `uint8`'s for static, uknown, and dynamic.

radarAaron replied:
done

tfoote commented on msg/RadarTrack.msg:
```diff
@@ -0,0 +1,23 @@
+geometry_msgs/Vector3 size                  # The object size as represented by the radar sensor eg. length, width, height OR the diameter of an ellipsoid in the x, y, z, dimensions
+                                            # and is from the sensor frame's view.
+uint8 classification                        # An optional classification of the object (see above)
+float64[9] position_covariance              # Row-major about the x, y, z axes
```
This is the current standard for covariance matrices [from REP 103](https://www.ros.org/reps/rep-0103.html) but it is very large. 

In the message API review it was deemed that a proposal to cut these down to lower resolution and upper triangular to cut the data size would be valuable: https://github.com/ros2/common_interfaces/issues/91 

Would people be interested in trying that out here. It's likely we can cut the covariance data size by two thirds just dropping to 32bit and upper triangular.

In this case it will save 192 of 288 bytes of covariance and will make this message overall about half the size per track. Since the current one has ~288 bytes of covariance and ~104 of position and uuid and one more for the classification. 

This would be a good amount of bandwidth savings.

SteveMacenski replied:
I'll yield to what you think is best on this front. The only criteria I think is important is consistency, so if other standard interfaces are going to move to this relatively soon, we can do that here too. If its not necessarily sure when that change would be made, I'd rather stick with tried and true.

radarAaron replied:
I have updated the covariance fields to float32's. I couldn't find any examples to copy about the wording to use for the upper triangle covariance matrix. Please review and suggest changes if required.

tfoote reviewed (commented):
Thanks for iterating this looks a lot stronger to me. It's ok as it is, but I have a few suggestions to consider for improving.

aaronfultonnz committed:
Altered description and data type of classification and changed the covariance fields

aaronfultonnz committed:
Updated types for object classification enums

JWhitleyWork reviewed (changes requested):
I just noticed that `geometry_msgs` and `uuid_msgs` are new dependencies for this repository so they need to be added to `package.xml` and `CMakeLists.txt` in this pull request.

aaronfultonnz committed:
Merge branch 'master' of https://github.com/ros-perception/radar_msgs into RadarTrack

aaronfultonnz committed:
Added dependencies to CMakelists.txt and package.xml

radarAaron commented:
good catch @JWhitleyWork changes made.

JWhitleyWork commented:
@SteveMacenski Can you and/or @tfoote address the one remaining item about the covariance fields? I think we're then ready for a merge!

SteveMacenski commented:
Waiting on @tfoote to respond, I re-asked for a review to have him take a look. I'm happy as is, but I want to make sure everyone is OK with this since its a standard that impacts more than just I

JWhitleyWork commented:
Ping to @tfoote.

icolwell-as commented:
Once this PR is merged we will be able to a noetic release right? Currently the lack of radar_msgs package in noetic is blocking us from releasing our radar driver. We aren't in any rush, just wanted to double-check that this is the last piece of the puzzle.

Thanks again for putting theses messages together in a standardized way!

JWhitleyWork commented:
@SteveMacenski Given that it's literally been a couple of months and the fact that this message hasn't even been used and so is likely to have some modification in the future anyway, I vote for going ahead with the merge for now and making updates as required in the future. Better to get it out there and have people use it (including Autoware.Auto) than wait for it to be perfect.

SteveMacenski commented:
Lets give it another week, but then yeah I agree. I would like to let @tfoote have another go at it since he's our supreme interface expert.

JWhitleyWork commented:
@SteveMacenski Ping for merge.

SteveMacenski commented:
Fair enough - I welcome to continue this discussion if there are additional requirements not present here, but I agree its time to ship it

SteveMacenski commented:
@radarAaron thanks for baring with us after these many months! Your work and others in this thread have made a great impact here. Thanks to all!
